util/promlint/promlint.go

```diff
@@ -12,17 +12,14 @@
 // limitations under the License.
 
 // Package promlint provides a linter for Prometheus metrics.
+//
+// Deprecated: Use the "github.com/prometheus/client_golang/prometheus/testutil/promlint" package instead.
 package promlint
 
 import (
-	"fmt"
 	"io"
-	"regexp"
-	"sort"
-	"strings"
 
-	dto "github.com/prometheus/client_model/go"
-	"github.com/prometheus/common/expfmt"
+	"github.com/prometheus/client_golang/prometheus/testutil/promlint"
```

Comment:
Can we not delete this entire file?

Reply:
My assuption was that we would replace  github.com/prometheus/prometheus/testutil/promlint by github.com/prometheus/client_golang/prometheus/testutil/promlint in our codebase indeed.

Reply:
> Can we not delete this entire file?

Do you mean remove the package directly? My concern is it will violate the compatibility practice.

Reply:
Yes, remove the entire package. There's no compatibility guarantees in this repository at the code level.

Reply:
To be honest, it's a shock to me.
But, I'm ok, if there are no compatibility guarantees.
cc @beorn7 @brancz

Reply:
I agree with brian, that is why we moved to client_golang, to provide those guarantees

Reply:
@RainbowMango so that's the problem with what semantic versioning refers to. The release versions for prometheus/prometheus refer to the behavior of the Prometheus binary (and promtool), not to code-level interfaces. That's different in prometheus/client_golang, which is a true library. Of course, many people use packages from prometheus/prometheus as libraries, and Go modules just assume the version tags in prometheus/prometheus refer to code level compatibility, which is simply not the case.

We discussed the problem a lot. There are several possible ways out of it. We haven't really settled on any of them yet, we haven't even decided if that's a problem that should be tackled at all, but https://github.com/modularise/modularise is a quite elaborate approach which also offers extracted versions of some prometheus/prometheus packages typically used externally:
- https://github.com/modularise/prometheus-tsdb
- https://github.com/modularise/prometheus-promql
- https://github.com/modularise/prometheus-discovery

Reply:
Copy that! Thanks for your information.